### PR TITLE
Split camera images

### LINKED ISSUE #116: Split camera images and distribute them as a virtual source

### PR DESCRIPTION
**Pull request by g3force**
Resolves #116 

 * Add Distributor stack that captures an image and distributes it to the splitters
 * Add Splitter capture which cuts a part of the images from the distributor and applies the default vision stacks on the resulting image
 * Reduce the buffer size for capture to visualization transmission to reduce latency, always read the latest element from buffer
 * Add CLI parameter to set the number of camera threads

### CODE REVIEW

File: src/app/gui/mainwindow.cpp [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
 #include "mainwindow.h"
 
-MainWindow::MainWindow(bool start_capture, bool enforce_affinity)
+MainWindow::MainWindow(bool start_capture, bool enforce_affinity, int num_cameras)
```
Comment (joydeep-b) [inline src/app/gui/mainwindow.cpp]: :+1: on making this a CLI parameter! :-)

File: src/app/stacks/multistack_robocup_ssl.cpp [hunk 2]
```diff
@@ -84,14 +86,16 @@ MultiStackRoboCupSSL::MultiStackRoboCupSSL(RenderOptions * _opts, int cameras) :
...
+  createThreads(num_threads);
+  std::vector<CaptureSplitter*> captureSplitters;
+  captureSplitters.resize((unsigned long) num_normal_camera_threads);
+  for (unsigned int i = 0; i < num_normal_camera_threads;i++) {
     //NOTE: if modified to put different stacks in cameras, please
     //      update the plugin_colorcalib.cpp code to safely reallocate and copy their
     //      data instead of assuming that format and size is uniform across
     //      cameras -- added when LUTs became aware of other cameras (Zavesky, 2/16)
-    threads[i]->setFrameBuffer(new FrameBuffer(5));
+    threads[i]->setFrameBuffer(new FrameBuffer(3));
```
Comment (joydeep-b) [inline src/app/stacks/multistack_robocup_ssl.cpp]: Why 3?

File: src/app/stacks/multivisionstack.cpp [hunk 3]
```diff
@@ -48,7 +48,7 @@ string MultiVisionStack::getSettingsFileName() {
 
 void MultiVisionStack::createThreads(int number) {
   for (int i=0;i<number;i++) {
-    threads.push_back(new CaptureThread(i));
+    threads.push_back(new CaptureThread(i%4));
```
Comment (joydeep-b) [inline src/app/stacks/multivisionstack.cpp]: Should this be `num_cameras` instead of `4`?

File: src/app/gui/mainwindow.cpp (hunk 1 above)
Comment (g3force) [reply src/app/gui/mainwindow.cpp]: Have you read my comment on the old PR? https://github.com/RoboCup-SSL/ssl-vision/pull/117#discussion_r259627279

File: src/app/stacks/multistack_robocup_ssl.cpp (hunk 2 above)
Comment (g3force) [reply src/app/stacks/multistack_robocup_ssl.cpp]: That's the minimum. With less, the application will block.

File: src/app/stacks/multivisionstack.cpp (hunk 3 above)
Comment (g3force) [reply src/app/stacks/multivisionstack.cpp]: Yeah, that should also work. Definitely looks better than the hard coded number :)

File: src/app/gui/mainwindow.cpp (hunk 1 above)
Comment (joydeep-b) [reply src/app/gui/mainwindow.cpp]: I see. Okay, that's fine for now.

**Pull request merged**